src/ray/gcs/pubsub/gcs_pub_sub.cc

```diff
@@ -291,5 +295,203 @@ Status PythonGcsPublisher::PublishFunctionKey(
+  auto *cmd = request.add_commands();
+  cmd->set_channel_type(channel_type_);
+  cmd->mutable_subscribe_message();
+
+  rpc::GcsSubscriberCommandBatchReply reply;
+  grpc::Status status =
+      pubsub_stub_->GcsSubscriberCommandBatch(&context, request, &reply);
+
+  if (status.ok()) {
+    return Status::OK();
+  } else {
+    return Status::RpcError(status.error_message(), status.error_code());
+  }
+}
+
+Status PythonGcsSubscriber::DoPoll(int64_t timeout_ms, rpc::PubMessage *message) {
+  absl::MutexLock lock(&mu_);
+
+  while (queue_.size() == 0) {
+    if (closed_) {
+      return Status::OK();
+    }
+    current_polling_context_.reset(new grpc::ClientContext());
+    if (timeout_ms != -1) {
+      current_polling_context_->set_deadline(std::chrono::system_clock::now() +
+                                             std::chrono::milliseconds(timeout_ms));
+    }
+    rpc::GcsSubscriberPollRequest request;
+    request.set_subscriber_id(subscriber_id_);
+    request.set_max_processed_sequence_id(max_processed_sequence_id_);
+    request.set_publisher_id(publisher_id_);
+
+    rpc::GcsSubscriberPollReply reply;
+    auto context = current_polling_context_;
+    // Drop the lock while in RPC
+    mu_.Unlock();
+    grpc::Status status = pubsub_stub_->GcsSubscriberPoll(context.get(), request, &reply);
+    mu_.Lock();
+
+    if (status.error_code() == grpc::StatusCode::DEADLINE_EXCEEDED ||
+        status.error_code() == grpc::StatusCode::UNAVAILABLE) {
+      return Status::OK();
+    } else if (status.error_code() == grpc::StatusCode::CANCELLED) {
+      // This channel was shut down via Close()
+      return Status::OK();
+    } else if (status.error_code() != grpc::StatusCode::OK) {
+      return Status::Invalid(status.error_message());
+    }
+
+    if (publisher_id_ != reply.publisher_id()) {
+      if (publisher_id_ != "") {
+        RAY_LOG(DEBUG) << "Replied publisher_id " << reply.publisher_id()
+                       << " different from " << publisher_id_
+                       << ", this should only happen"
+                       << " during GCS failover.";
+      }
+      publisher_id_ = reply.publisher_id();
+      max_processed_sequence_id_ = 0;
+    }
+    last_batch_size_ = reply.pub_messages().size();
+    for (auto &message : reply.pub_messages()) {
+      if (message.sequence_id() <= max_processed_sequence_id_) {
+        RAY_LOG(WARNING) << "Ignoring out of order message " << message.sequence_id();
+        continue;
+      }
+      max_processed_sequence_id_ = message.sequence_id();
+      if (message.channel_type() != channel_type_) {
+        RAY_LOG(WARNING) << "Ignoring message from unsubscribed channel "
+                         << message.channel_type();
+        continue;
+      }
+      queue_.emplace_back(std::move(message));
+    }
+  }
+
+  *message = queue_.front();
+  queue_.pop_front();
+
+  return Status::OK();
+}
+
+Status PythonGcsSubscriber::PollError(std::string *key_id,
```

Comment:
Purely a suggestion because semantics are ultimately identical. Instead of having all this boilerplate, we could template by data and do a compile-time check for setting the correct oneof field
```
template class <RPCData>
Status PythonGcsSubscriber::Poll(std::string *key_id, int64_t timeout_ms, RPCData *data) {
  rpc::PubMessage message;
  RAY_RETURN_NOT_OK(DoPoll(timeout_ms, &message));
  *key_id = message.key_id();
  if constexpr (std::is_same<RPCData, rpc::ErrorTableData>::value) {
    *data = message.error_info_message();
  } else if constexpr (...) {
  ...
  } else {
    RAY_CHECK(false) << "Not a valid polling type.";
  }
  return Status::OK();
}
```

Reply:
That's neat, but I don't think Cython can wrap those templates :|

And if we write all the wrappers explicitly, it is pretty similar to what is there now (in that case I would rather not introduce this templating).